### ISSUE TITLE
find the version range for base

### ISSUE THREAD

**Issue by gelisam** [gelisam/cabal-rangefinder#2]
The "base" package is determined by the version of GHC which is installed on the system, I can't just install every single version of the Haskell platform in order to test which ones work. Well, actually, I could; I just don't know how to install them in a sandboxed way, that is, in a way which causes my package to use that specific version of GHC, without causing anything else on the machine to use that version of GHC.

Brew on OS X and GoboLinux have an interesting symlink-based installation system which allows multiple versions of a package to be installed on the same machine. I could probably ask the system to install all the versions of GHC and then switch between the different versions, making sure to revert to the original version on exit. But is there a more portable solution?

**Comment by Wizek** [comment]
Have you heard about [`stack`](https://github.com/commercialhaskell/stack)?

If not it is intended as a `cabal-install` replacement by its authors, but that's beside my main point.

It is capable to install and manage multiple GHC versions in the background without touching the global installation and switch to one or the other for building when running `stack setup`.

Maybe there can be some inspiration drawn on how that tool does it, and `cabal-rangefinder` could do the same?

**Comment by gelisam** [comment]
> Have you heard about `stack`?

Of course! It's a new tool which didn't exist back when this issue was created :)

> Maybe there can be some inspiration drawn on how that tool does it, and `cabal-rangefinder` could do the same?

That's a good idea! Although, I'd rather not duplicate existing code, so I think a better approach would be to add `stack` as a dependency of `cabal-rangefinder` and to use it to install different versions of GHC. In fact I think I might also be able to use `stack` to perform the compilation as well: I should be able to use Cabal to find an install plan, and then ask `stack` to build this exact set of package versions. This should be faster, since `stack` somehow manages not to have to clean its sandbox between every installation. Although... that's probably because they use stackage to pin the versions of every package involved, so for my use case, not cleaning the sandbox might not be safe.

Indeed, here's what I found in their [architecture](https://github.com/commercialhaskell/stack/wiki/Architecture) page:

> If different projects use different package indices, then their definitions of what package foo-1.2.3 are may be different, in which case they _can_ corrupt each other's shared databases. This is warned about in the FAQ

Their [FAQ](https://github.com/commercialhaskell/stack/wiki/FAQ) page does not, in fact, warn about this. Maybe they since found a fix for this and forgot to update the architecture page?

**Comment by gelisam** [comment]
Taking a closer reading, "different package indices" is not a weird way to talk about `foo-1.0` vs `foo-2.0`, instead "package indices" is the plural of "package index", such as Hackage or Stackage.

So using stack to do all the compilation should work just fine! Yay!
